index/index.go

```diff
@@ -0,0 +1,498 @@
+		ioProvider:  ioProvider,
+		idProvider:  idProvider,
+		indexKey:    keys.IEK,
+	}, nil
+}
+
+// PlainID contains a document ID and the counter used to compute the next sealed ID.
+type PlainID struct {
+	DocID       string
+	NextCounter uint64
+}
+
+// SealedID is an encrypted structure which defines an occurrence of a specific keyword in a specific DocID.
+type SealedID struct {
+	Ciphertext []byte
+	WrappedKey []byte
+}
+
+// Seal encrypts the plaintext ID.
+func (i *PlainID) Seal(label uuid.UUID, cryptor crypto.CryptorInterface) (SealedID, error) {
+	wrappedKey, ciphertext, err := cryptor.Encrypt(i, label.Bytes())
+	if err != nil {
+		return SealedID{}, err
+	}
+
+	sealed := SealedID{
+		Ciphertext: ciphertext,
+		WrappedKey: wrappedKey,
+	}
+
+	return sealed, nil
+}
+
+// Unseal decrypts the sealed ID.
+func (i *SealedID) Unseal(label uuid.UUID, cryptor crypto.CryptorInterface) (PlainID, error) {
+	plainID := PlainID{}
+	if err := cryptor.Decrypt(&plainID, label.Bytes(), i.WrappedKey, i.Ciphertext); err != nil {
+		return PlainID{}, err
+	}
+	return plainID, nil
+}
+
+// Add adds a keyword/docID pair to the secure index.
+func (i *SecureIndex) Add(key []byte, token, keyword, docID string) error {
+	identity, err := i.idProvider.GetIdentity(token)
+	if err != nil {
+		return ErrNotAuthenticated
+	}
+	if !identity.Scopes.Contains(id.ScopeIndex) {
+		return ErrNotAuthorized
+	}
+
+	if len(key) != MasterKeyLength {
+		return ErrInvalidMasterKeyLength
+	}
+
+	// Create two keys, k1 and k2: One for tagger and one for cryptor.
+	k1 := crypto.KMACKDF(crypto.TaggerKeyLength, key, []byte("label"), []byte(keyword))
+	k2 := crypto.KMACKDF(crypto.EncryptionKeyLength, key, []byte("id"), []byte(keyword))
+
+	tagger, err := crypto.NewKMAC256Tagger(k1)
+	if err != nil {
+		return err
+	}
+	cryptor, err := crypto.NewAESCryptor(k2)
+	if err != nil {
+		return err
+	}
+
+	// Compute the current last sealed ID with the given keyword, i.e. the sealed ID with
+	// the largest value of the counter.
+	lastCounter, lastID, err := i.lastID(key, keyword)
+	if err != nil {
+		return err
+	}
+
+	counter := uint64(0)
+
+	// If the current last sealed ID, lastID, is not an empty sealed ID,
+	// i.e. the keyword has already been added to the secure index, do the following.
+	if lastID.DocID != "" {
+		// Compute the next counter.
+		counter = lastCounter + 1
+
+		// Update lastID's NextCounter.
+		lastLabelUUID, err := computeLabelUUID(lastCounter, &tagger)
+		if err != nil {
+			return err
+		}
+
+		lastID.NextCounter = counter
+
+		lastSealedID, err := lastID.Seal(lastLabelUUID, &cryptor)
+		if err != nil {
+			return err
+		}
+		if err = i.putSealedID(lastLabelUUID, &lastSealedID, true); err != nil {
+			return err
+		}
+	}
+
+	// Compute new labelUUID and plaintext ID, seal it, and send it to the IO Provider.
+	labelUUID, err := computeLabelUUID(counter, &tagger)
+	if err != nil {
+		return err
+	}
+
+	plainID := PlainID{DocID: docID, NextCounter: 0}
```

Comment:
I think we can significantly simplify a lot of the logic by having `NextCounter` always point to something (even if it doesn't exist yet). Let's discuss.